### PR TITLE
TASK: Incorporate AFX parser into Neos.Fusion.Afx

### PR DESCRIPTION
**Pull request by mficzel**
The external library packagefactory/afx is incorporated into the fusion-afx package for beeing developed in unison in future.

### CODE REVIEW

File: Classes/Parser/Expression/Prop.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+                case $lexer->isOpeningBrace():
+                    $value = [
+                        'type' => 'expression',
+                        'payload' => Expression::parse($lexer),
+                        'identifier' => $identifier
+                    ];
+                    break;
+                default:
+                    throw new Exception(sprintf(
```
Comment (daniellienert) [inline Classes/Parser/Expression/Prop.php]: Would you please add an exception code to the exception.

File: Classes/Parser/Expression/Prop.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                    ));
+            }
+        } elseif ($lexer->isWhiteSpace() || $lexer->isForwardSlash() || $lexer->isClosingBracket()) {
+            $value = [
+                'type' => 'boolean',
+                'payload' => true,
+                'identifier' => $identifier
+            ];
+        } else {
+            throw new Exception(sprintf('Prop identifier "%s" is neither assignment nor boolean', $identifier));
```
Comment (daniellienert) [inline Classes/Parser/Expression/Prop.php]: Would you please add an exception code to the exception.

File: Classes/Parser/Expression/Spread.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            $lexer->consume();
+            $lexer->consume();
+            $lexer->consume();
+        } else {
+            throw new Exception('Spread without braces');
+        }
+
+        while (true) {
+            if ($lexer->isEnd()) {
+                throw new Exception(sprintf('Unfinished Spread "%s"', $contents));
```
Comment (daniellienert) [inline Classes/Parser/Expression/Spread.php]: Would you please add an exception code to the exception.

File: Classes/Parser/Expression/Spread.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        $contents = '';
+        $braceCount = 0;
+
+        if ($lexer->isOpeningBrace() && $lexer->peek(4) === '{...') {
+            $lexer->consume();
+            $lexer->consume();
+            $lexer->consume();
+            $lexer->consume();
+        } else {
+            throw new Exception('Spread without braces');
```
Comment (daniellienert) [inline Classes/Parser/Expression/Spread.php]: Would you please add an exception code to the exception.

File: Classes/Parser/Parser.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+class Parser
+{
+    public function __construct($string)
+    {
+        $this->lexer = new Lexer($string);
```
Comment (daniellienert) [inline Classes/Parser/Parser.php]: The property lexer is not defined here.

File: composer.json [hunk 6]
```diff
@@ -7,8 +7,7 @@
     "packagefactory/atomicfusion-afx": "*"
   },
   "require": {
-    "neos/fusion": "^4.2.0 || dev-master",
-    "packagefactory/afx": "^3.0.0"
+    "neos/fusion": "^4.2.0 || dev-master"
   },
   "require-dev": {
     "phpunit/phpunit": "~6.0.0"
```
Comment (daniellienert) [inline composer.json]: Would you please raise phpunit to "^8.1" which we now use in all packages.

File: Classes/Parser/Expression/Expression.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public static function parse(Lexer $lexer): string
+    {
+        $contents = '';
+        $braceCount = 0;
+
+        if ($lexer->isOpeningBrace()) {
+            $lexer->consume();
+        } else {
+            throw new AfxParserException('Expression without braces', 1557860467921);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Expression.php]:
```suggestion
            throw new AfxParserException('Expression without braces', 1557860467);
```

File: Classes/Parser/Expression/Expression.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        while (true) {
+            if ($lexer->isEnd()) {
+                throw new AfxParserException(sprintf('Unfinished Expression "%s"', $contents), 1557860496139);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Expression.php]:
```suggestion
                throw new AfxParserException(sprintf('Unfinished Expression "%s"', $contents), 1557860496);
```

File: Classes/Parser/Expression/Node.php [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                    $lexer->consume();
+
+                    return [
+                        'identifier' => $identifier,
+                        'attributes' => $attributes,
+                        'children' => $children,
+                        'selfClosing' => true
+                    ];
+                } else {
+                    throw new AfxParserException(sprintf('Self closing tag "%s" misses closing bracket.', $identifier), 1557860567682);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Node.php]:
```suggestion
                    throw new AfxParserException(sprintf('Self closing tag "%s" misses closing bracket.', $identifier), 1557860567);
```

File: Classes/Parser/Expression/Node.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                }
+            }
+
+            if ($lexer->isClosingBracket()) {
+                $lexer->consume();
+            } else {
+                throw new AfxParserException(sprintf('Tag "%s" did not end with closing bracket.', $identifier), 1557860573945);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Node.php]:
```suggestion
                throw new AfxParserException(sprintf('Tag "%s" did not end with closing bracket.', $identifier), 1557860573);
```

File: Classes/Parser/Expression/Node.php [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                    'attributes' => $attributes,
+                    'children' => $children,
+                    'selfClosing' => false
+                ];
+            } else {
+                throw new AfxParserException(sprintf('Closing tag "%s" did not end with closing-bracket.', $identifier), 1557860618559);
+            }
+
+            if ($lexer->isEnd()) {
+                throw new AfxParserException(sprintf('Tag was %s is not closed.', $identifier), 1557860622684);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Node.php]:
```suggestion
                throw new AfxParserException(sprintf('Tag was %s is not closed.', $identifier), 1557860622);
```

File: Classes/Parser/Expression/Node.php [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            }
+        } catch (AfxParserException $e) {
+            throw new AfxParserException(sprintf('<%s> %s', $identifier, $e->getMessage()), 1557860627201);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Node.php]:
```suggestion
            throw new AfxParserException(sprintf('<%s> %s', $identifier, $e->getMessage()), 1557860627);
```

File: Classes/Parser/Expression/Prop.php [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+                    ), 1557860545099);
+            }
+        } elseif ($lexer->isWhiteSpace() || $lexer->isForwardSlash() || $lexer->isClosingBracket()) {
+            $value = [
+                'type' => 'boolean',
+                'payload' => true,
+                'identifier' => $identifier
+            ];
+        } else {
+            throw new AfxParserException(sprintf('Prop identifier "%s" is neither assignment nor boolean', $identifier), 1557860552278);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Prop.php]:
```suggestion
            throw new AfxParserException(sprintf('Prop identifier "%s" is neither assignment nor boolean', $identifier), 1557860552);
```

File: Classes/Parser/Expression/Spread.php [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        $contents = '';
+        $braceCount = 0;
+
+        if ($lexer->isOpeningBrace() && $lexer->peek(4) === '{...') {
+            $lexer->consume();
+            $lexer->consume();
+            $lexer->consume();
+            $lexer->consume();
+        } else {
+            throw new AfxParserException('Spread without braces', 1557860522837);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Spread.php]:
```suggestion
            throw new AfxParserException('Spread without braces', 1557860522);
```

File: Classes/Parser/Expression/Spread.php [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        while (true) {
+            if ($lexer->isEnd()) {
+                throw new AfxParserException(sprintf('Unfinished Spread "%s"', $contents), 1557860526973);
```
Comment (daniellienert) [inline Classes/Parser/Expression/Spread.php]:
```suggestion
                throw new AfxParserException(sprintf('Unfinished Spread "%s"', $contents), 1557860526);
```

File: Classes/Parser/Expression/StringLiteral.php [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public static function parse(Lexer $lexer): string
+    {
+        $openingQuoteSign = '';
+        $contents = '';
+        $willBeEscaped = false;
+        if ($lexer->isSingleQuote() || $lexer->isDoubleQuote()) {
+            $openingQuoteSign = $lexer->consume();
+        } else {
+            throw new AfxParserException('Unquoted String literal', 1557860514707);
```
Comment (daniellienert) [inline Classes/Parser/Expression/StringLiteral.php]:
```suggestion
            throw new AfxParserException('Unquoted String literal', 1557860514);
```

File: Classes/Parser/Expression/StringLiteral.php [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        while (true) {
+            if ($lexer->isEnd()) {
+                throw new AfxParserException(sprintf('Unfinished string literal "%s"', $contents), 1557860504068);
```
Comment (daniellienert) [inline Classes/Parser/Expression/StringLiteral.php]:
```suggestion
                throw new AfxParserException(sprintf('Unfinished string literal "%s"', $contents), 1557860504);
```

File: Classes/Parser/Expression/Identifier.php [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+                case $lexer->isForwardSlash():
+                    return $identifier;
+                    break;
+                default:
+                    $unexpected_character = $lexer->consume();
+                    throw new AfxParserException(sprintf(
+                        'Unexpected character "%s" in identifier "%s"',
+                        $unexpected_character,
+                        $identifier
+                    ), 1557860650835);
```
Comment (Sebobo) [inline Classes/Parser/Expression/Identifier.php]:
```suggestion
                    ), 1557860650);
```

File: Classes/Parser/Expression/Node.php [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+            if ($lexer->isOpeningBracket()) {
+                $lexer->consume();
+
+                if ($lexer->isForwardSlash()) {
+                    $lexer->consume();
+                } else {
+                    throw new AfxParserException(sprintf(
+                        'Opening-bracket for closing of tag "%s" was not followed by slash.',
+                        $identifier
+                    ), 1557860584196);
```
Comment (Sebobo) [inline Classes/Parser/Expression/Node.php]:
```suggestion
                    ), 1557860584);
```

File: Classes/Parser/Expression/Node.php [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+                }
+            } else {
+                throw new AfxParserException(sprintf(
+                    'Opening-bracket for closing of tag "%s" expected.',
+                    $identifier
+                ), 1557860587680);
```
Comment (Sebobo) [inline Classes/Parser/Expression/Node.php]:
```suggestion
                ), 1557860587);
```

File: Classes/Parser/Expression/Node.php [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+            }
+
+            $closingIdentifier = Identifier::parse($lexer);
+
+            if ($closingIdentifier !== $identifier) {
+                throw new AfxParserException(sprintf(
+                    'Closing-tag identifier "%s" did not match opening-tag identifier "%s".',
+                    $closingIdentifier,
+                    $identifier
+                ), 1557860595281);
```
Comment (Sebobo) [inline Classes/Parser/Expression/Node.php]:
```suggestion
                ), 1557860595);
```

File: Classes/Parser/Expression/Node.php [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+            if ($lexer->isClosingBracket()) {
+                $lexer->consume();
+                return [
+                    'identifier' => $identifier,
+                    'attributes' => $attributes,
+                    'children' => $children,
+                    'selfClosing' => false
+                ];
+            } else {
+                throw new AfxParserException(sprintf('Closing tag "%s" did not end with closing-bracket.', $identifier), 1557860618559);
```
Comment (Sebobo) [inline Classes/Parser/Expression/Node.php]:
```suggestion
                throw new AfxParserException(sprintf('Closing tag "%s" did not end with closing-bracket.', $identifier), 1557860618);
```

File: Classes/Parser/Expression/Prop.php [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+                        'type' => 'expression',
+                        'payload' => Expression::parse($lexer),
+                        'identifier' => $identifier
+                    ];
+                    break;
+                default:
+                    throw new AfxParserException(sprintf(
+                        'Prop-assignment "%s" was not followed by quotes or braces',
+                        $identifier
+                    ), 1557860545099);
```
Comment (Sebobo) [inline Classes/Parser/Expression/Prop.php]:
```suggestion
                    ), 1557860545);
```

**Pull request merged**